workspaces/diff-engine/src/learn_shape/result.rs

```diff
@@ -56,13 +60,19 @@ impl TrailObservationsResult {
 
     let mut shape_prototypes_by_trail = HashMap::new();
     let mut shape_prototypes = Vec::with_capacity(sorted_trails.len());
+    let mut new_shape_id: Option<String> = None;
 
     for json_trail in sorted_trails.into_iter().rev() {
       let trail_values = self.values_by_trail.remove(&json_trail).unwrap();
 
       let shape_prototype =
         trail_values.into_shape_prototype(id_generator, &shape_prototypes_by_trail);
 
+      // collect the id of the first shape prototype we create
+      if new_shape_id.is_none() {
```

Comment:
@devdoshi is there a better Rust way of doing this?

Reply:
we preallocate the vec of shape_prototypes so you could clone it from the last vec element after the iteration is done, instead of checking it every iteration

Reply:
Also, should it be the first or the last?